### ISSUE TITLE
Tracking epoch mismatch in remove_abandoned

### ISSUE THREAD

**Issue by kvark** [gfx-rs/wgpu#830]
Get an assertion when trying to reproduce #828:
```
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `4`,
 right: `3`', /Users/dmalyshau/Code/wgpu/wgpu-core/src/track/mod.rs:218:21
```
stack:
```
  13: wgpu_core::track::ResourceTracker<S>::remove_abandoned
             at /Users/dmalyshau/.cargo/git/checkouts/wgpu-rs-40ea39809c03c5d8/0c7c3af/<::std::macros::panic macros>:9
  14: wgpu_core::device::life::LifetimeTracker<B>::triage_suspected
             at /Users/dmalyshau/Code/wgpu/wgpu-core/src/device/life.rs:403
  15: wgpu_core::device::Device<B>::maintain
             at /Users/dmalyshau/Code/wgpu/wgpu-core/src/device/mod.rs:303
  16: wgpu_core::device::queue::<impl wgpu_core::hub::Global<G>>::queue_submit
             at /Users/dmalyshau/Code/wgpu/wgpu-core/src/device/queue.rs:536
  17: wgpu::backend::direct::<impl wgpu::Context for wgpu_core::hub::Global<wgpu_core::hub::IdentityManagerFactory>>::queue_submit
```

**Comment by kvark** [comment]
Here is the fun thing that happens here:
  1. the same texture ID is pushed multiple times into the suspected list. This is fine today (we didn't expect issues because of that), e.g. when multiple bind groups are removed, all pointing to the same texture.
  2. first ID is encountered, and we figure that the ID is used last, and should be removed
  3. we remove the texture and free the ID
  4. in parallel, on a different thread, a texture is created, and it gets that freshly freed ID with a new epoch
  5. the first thread continues going through suspected IDs, and it sees the other IDs of the same texture
  6. we are trying to remove it, and we see that the epoch is unexpected

The moral of the story is: in multi-threaded environments, the suspected list should really be a *set* instead, uniquely holding each key.

**Comment by kvark** [comment]
I celebrate the fact that we can create textures literally in the middle of a cleanup routine on the device 🥇 
Going to fix this promptly though to be correct ;)

**Comment by kvark** [comment]
Ok, that was wrong, actually. Need to look more into it 😓

**Comment by kvark** [comment]
Ok, here is the *real* explanation of events. When buffers or textures are destroyed, they don't instantly go into the "suspected" list. Instead, they go into the "future suspected" list that will be merged into "suspected" on the next submit. This is because there might be pending `write_*` operations on those resources, which we haven't even queued yet by the time they are deleted.

Problem occurred when they were *also* happen to be in the "suspected" list. Since we kill the refcount instantly, we'd actually remove them right away, and then try removing them again after the submit.
